File: Sources/SSAlertOC/SSAlertView.m
//
//  SSAlertCustomView.m
//  SSAlertView
//
//  Created by yangsq on 2020/1/13.
//  Copyright © 2020 yangsq. All rights reserved.
//

#import "SSAlertView.h"
#import "UIView+SS_Add.h"
#import "SSAlertViewAnimationController.h"
#import "SSAlertViewPresentAnimation.h"

@interface SSAlertView()
@property (nonatomic, weak) UIView *customView;
@property (nonatomic, weak) UIView *onView;
@property (nonatomic, strong) UIButton *backgroundMaskButton;
///遮罩类型
@property (nonatomic, assign) SSAlertBackgroundMaskType masktype;

@property (nonatomic, weak) UIViewController *fromViewController;

@property (nonatomic, weak) UINavigationController *navigationController;

@property (nonatomic, weak) UIViewController *toViewContrller;

@property (nonatomic, strong) SSAlertViewPresentAnimation *presentAnimation;

@property (nonatomic, assign) BOOL showPanDimissGestrue;

@property (nonatomic, copy) void(^hideCompletion)(void);


@end

@implementation SSAlertView

- (id)initWithCustomView:(UIView *)customView
                  onView:(nonnull UIView *)onView{
    return [self initWithCustomView:customView onView:onView masktype:SSAlertBackgroundMaskTypeNone];
}

- (id)initWithCustomView:(UIView *)customView
                  onView:(UIView *)onView
                masktype:(SSAlertBackgroundMaskType)masktype {
    if (self = [super initWithFrame:CGRectZero]) {
        if (onView) {
            self.customView = customView;
            self.onView = onView;
            self.masktype = masktype;
            [self setupViews];
        }
    }
    
    return self;
}



- (id)initWithCustomView:(UIView *)customView
      fromViewController:(UIViewController *)fromViewController
navigationControllerClass:(Class)navigationControllerClass
                masktype:(SSAlertBackgroundMaskType)masktype
{
    return [self initWithCustomView:customView fromViewController:fromViewController navigationControllerClass:navigationControllerClass masktype:masktype canPanDimiss:NO];
}

- (id)initWithCustomView:(UIView *)customView
      fromViewController:(UIViewController *)fromViewController
navigationControllerClass:(Class)navigationControllerClass
                masktype:(SSAlertBackgroundMaskType)masktype canPanDimiss:(BOOL)showPanDimissGestrue {
    if (self = [super initWithFrame:CGRectZero]) {
        if (fromViewController) {
            self.customView = customView;
            self.fromViewController = fromViewController;
            self.masktype = masktype;
            self.showPanDimissGestrue = showPanDimissGestrue;
            SSAlertViewAnimationController *animationVC = [[SSAlertViewAnimationController alloc]init];
            UIViewController * superVC = [self superViewController:customView];
            if (superVC != nil) {
                [animationVC addChildViewController:superVC];
                [superVC didMoveToParentViewController:animationVC];
            }
            UINavigationController *nav = [(UINavigationController *)[navigationControllerClass alloc]initWithRootViewController:animationVC];
            nav.modalPresentationStyle = UIModalPresentationCustom;
            self.navigationController = nav;
            SSAlertViewPresentAnimation *vc_animation = [[SSAlertViewPresentAnimation alloc]initWithNavigationViewController:self.navigationController animationView:customView showPanDimissGestrue:showPanDimissGestrue];
            self.navigationController.transitioningDelegate = vc_animation;
            self.navigationController.modalPresentationCapturesStatusBarAppearance = YES; //接管状态栏设置
            self.presentAnimation = vc_animation;
            self.toViewContrller = animationVC;
            self.onView = animationVC.view;
            [self setupViews];
           
        }
    }
    
    return self;
}

- (void)setupViews {
    self.canTouchMaskHide = YES;
    if (self.customView) {
        if (self.masktype != SSAlertBackgroundMaskTypeNone) {
            [self.onView addSubview:self.backgroundMaskButton];
            [self.backgroundMaskButton addTarget:self action:@selector(backgroundMaskAction) forControlEvents:UIControlEventTouchUpInside];
        }
        [self addSubview:self.customView];
        [self.onView addSubview:self];
    }
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.backgroundMaskButton.frame = CGRectMake(0, 0, self.onView.ss_w, self.onView.ss_h);
    self.customView.ss_x = self.customEdgeInsets.left;
    self.customView.ss_y = self.customEdgeInsets.top;
    self.customView.ss_w = CGRectGetWidth(self.frame) - self.customEdgeInsets.left - self.customEdgeInsets.right;
    self.customView.ss_h = CGRectGetHeight(self.frame) - self.customEdgeInsets.top - self.customEdgeInsets.bottom;
}

- (void)show:(BOOL)animated {
    if (self.navigationController != nil) {
        [self presentView:animated];
    } else {
        [self showView:animated];
    }
}
- (void)hide:(BOOL)animated {
    if (self.navigationController != nil) {
        [self dimissView:animated];
    } else {
        [self hideView:animated];
    }
}

- (void)showView:(BOOL)animation {
    
    if ([self.animation respondsToSelector:@selector(showAnimationOfAnimationView:viewSize:animation: completion:)]) {
        CGSize customViewSize = [self.customView systemLayoutSizeFittingSize:UILayoutFittingExpandedSize];
        if (!CGSizeEqualToSize(self.customView.ss_size, CGSizeZero)) {
            customViewSize = self.customView.ss_size;
        }
        CGSize size = CGSizeMake(customViewSize.width + self.customEdgeInsets.left + self.customEdgeInsets.right, customViewSize.height + self.customEdgeInsets.top + self.customEdgeInsets.bottom);
        [self.animation showAnimationOfAnimationView:self viewSize:size animation:animation completion:nil];
        
    }
    
}

- (void)hideView:(BOOL)animation {
    if ([self.animation respondsToSelector:@selector(hideAnimationOfAnimationView:viewSize:animation: completion:)]) {
        CGSize customViewSize = [self.customView systemLayoutSizeFittingSize:UILayoutFittingExpandedSize];
        if (!CGSizeEqualToSize(self.customView.ss_size, CGSizeZero)) {
            customViewSize = self.customView.ss_size;
        }
        CGSize size = CGSizeMake(customViewSize.width + self.customEdgeInsets.left + self.customEdgeInsets.right, customViewSize.height + self.customEdgeInsets.top + self.customEdgeInsets.bottom);
        [self.animation hideAnimationOfAnimationView:self viewSize:size animation:animation completion:^BOOL(BOOL finished) {
            return  NO;
        }];
        if (self.hideCompletion) {
            self.hideCompletion();
        }
    }
    

}

- (void)presentView:(BOOL)animation {
    
    if ([self.animation respondsToSelector:@selector(showAnimationOfAnimationView:viewSize:animation: completion:)]) {
        CGSize customViewSize = [self.customView systemLayoutSizeFittingSize:UILayoutFittingExpandedSize];
        if (!CGSizeEqualToSize(self.customView.ss_size, CGSizeZero)) {
            customViewSize = self.customView.ss_size;
        }
        CGSize size = CGSizeMake(customViewSize.width + self.customEdgeInsets.left + self.customEdgeInsets.right, customViewSize.height + self.customEdgeInsets.top + self.customEdgeInsets.bottom);
        __weak typeof(self) weakSelf = self;
        [self.presentAnimation setShowAnimation:^{
            __strong typeof (self) strongSelf = weakSelf;
            [strongSelf.animation showAnimationOfAnimationView:strongSelf viewSize:size animation:animation completion:^(BOOL finished) {
                [strongSelf.presentAnimation setCompleteTransitionIsHide:NO];
            }];
        }];
        self.presentAnimation.duration = [self.animation animationDuration];
        ///存在的问题： app当前显示的是弹窗界面，默认presnent出来的视图且modalPresentationStyle不是CurrentContext这这种，这个视图的presentingVC默认根控制器，例如存在tabbarVC的话就是tabbarVC，这个会造成一个问题：当这个视图dimiss的时候，会把所有的presentedVC都dimiss（其实调用dimiss的时候是调用当前视图的presentigngVC去dimiss,遵循随创建随释放）
        /// 这个通过设置 definesPresentationContext = true,modalPresentationStyle = UIModalPresentationOverCurrentContext，这样当前视图的presentigngVC为 self.navigationController,
        /// UIModalPresentationOverCurrentContext和 UIModalPresentationCurrentContext的区别是弹出的界面，前者不会把底部的视图stact移除，后者则会。
        self.navigationController.definesPresentationContext = true;
        self.navigationController.modalPresentationStyle = UIModalPresentationOverCurrentContext;
        [self.fromViewController presentViewController:self.navigationController animated:animation completion:nil];
        if (_showPanDimissGestrue) {
            [self dimissViewDragToHideAnimation: animation];
        }
    }
    
    
}

- (void)dimissView:(BOOL)animation {
    if ([self.animation respondsToSelector:@selector(hideAnimationOfAnimationView:viewSize:animation: completion:)]) {
          CGSize customViewSize = [self.customView systemLayoutSizeFittingSize:UILayoutFittingExpandedSize];
          if (!CGSizeEqualToSize(self.customView.ss_size, CGSizeZero)) {
              customViewSize = self.customView.ss_size;
          }
          CGSize size = CGSizeMake(customViewSize.width + self.customEdgeInsets.left + self.customEdgeInsets.right, customViewSize.height + self.customEdgeInsets.top + self.customEdgeInsets.bottom);
        __weak typeof(self) weakSelf = self;
        [self.presentAnimation setHideAnimation:^{
            __strong typeof (self) strongSelf = weakSelf;
            [strongSelf.animation hideAnimationOfAnimationView:strongSelf viewSize:size animation:animation completion:^BOOL(BOOL finished) {
                return  [strongSelf.presentAnimation setCompleteTransitionIsHide:YES];
            }];
        }];
        self.presentAnimation.duration = [self.animation animationDuration];
        [self.toViewContrller dismissViewControllerAnimated:YES completion:^{
            __strong typeof (self) strongSelf = weakSelf;
            if (strongSelf.hideCompletion) {
                strongSelf.hideCompletion();
            }
        }];
    
         
      }
}

- (void)dimissViewDragToHideAnimation:(BOOL)animation {
    if ([self.animation respondsToSelector:@selector(hideAnimationOfAnimationView:viewSize:animation: completion:)]) {
          CGSize customViewSize = [self.customView systemLayoutSizeFittingSize:UILayoutFittingExpandedSize];
          if (!CGSizeEqualToSize(self.customView.ss_size, CGSizeZero)) {
              customViewSize = self.customView.ss_size;
          }
          CGSize size = CGSizeMake(customViewSize.width + self.customEdgeInsets.left + self.customEdgeInsets.right, customViewSize.height + self.customEdgeInsets.top + self.customEdgeInsets.bottom);
        __weak typeof(self) weakSelf = self;
        [self.presentAnimation setHideAnimation:^{
            __strong typeof (self) strongSelf = weakSelf;
            [strongSelf.animation hideAnimationOfAnimationView:strongSelf viewSize:size animation:animation completion:^BOOL(BOOL finished) {
                return  [strongSelf.presentAnimation setCompleteTransitionIsHide:YES];
            }];
        }];
        self.presentAnimation.duration = [self.animation animationDuration];
        [self.presentAnimation setEndCompletedAnimation:^{
            __strong typeof (self) strongSelf = weakSelf;
            if (strongSelf.hideCompletion) {
                strongSelf.hideCompletion();
            }
        }];
        [self.presentAnimation setPanDimissAnimation:^CGFloat(CGPoint point, CGRect panViewFrame) {
            __strong typeof (self) strongSelf = weakSelf;
            return [strongSelf.animation panToDimissTransilatePoint:point panViewFrame:panViewFrame];
        }];
      }
}

- (void)refreshFrame:(BOOL)animation {
    CGSize customViewSize = [self.customView systemLayoutSizeFittingSize:UILayoutFittingExpandedSize];
    if (!CGSizeEqualToSize(self.customView.ss_size, CGSizeZero)) {
        customViewSize = self.customView.ss_size;
    }
    CGSize size = CGSizeMake(customViewSize.width + self.customEdgeInsets.left + self.customEdgeInsets.right, customViewSize.height + self.customEdgeInsets.top + self.customEdgeInsets.bottom);
    [self.animation refreshAnimationOfAnimationView:self viewSize:size animation:animation completion:nil];
}


- (void)backgroundMaskAction {
    if (self.canTouchMaskHide) {
        if (self.fromViewController) {
            [self dimissView:YES];
        }else {
            [self hideView:YES];
        }
    }
}

- (UIViewController *)superViewController:(UIView *)view {
    UIResponder * next = view.nextResponder;
    while (next != nil) {
        if ([next isKindOfClass: [UIViewController class]]) {
            return (UIViewController *)next;
        } else {
            next = next.nextResponder;
        }
    }
    return  nil;
}

- (void)observeHideCompletion:(void (^)(void))hideCompletion {
    self.hideCompletion = hideCompletion;
}


#pragma mark - getter


- (UIButton *)backgroundMaskButton {
    if (!_backgroundMaskButton) {
        _backgroundMaskButton = [UIButton buttonWithType:UIButtonTypeCustom];
        _backgroundMaskButton.alpha = 0;
    }
    
    return _backgroundMaskButton;
}

- (UIView *)backgroundMask {
    if (self.masktype == SSAlertBackgroundMaskTypeNone) {
        return  nil;
    }
    return self.backgroundMaskButton;
}


#pragma mark - setter

- (void)setCustomEdgeInsets:(UIEdgeInsets)customEdgeInsets {
    _customEdgeInsets = customEdgeInsets;
    [self setNeedsLayout];
}

- (void)setMasktype:(SSAlertBackgroundMaskType)masktype {
    _masktype = masktype;
    if (masktype != SSAlertBackgroundMaskTypeNone) {
        if (masktype == SSAlertBackgroundMaskTypeBlack) {
            self.backgroundMaskButton.backgroundColor = [[UIColor blackColor]colorWithAlphaComponent:0.3];
        }
    }
}


- (void)dealloc
{
    self.navigationController = nil;
    self.toViewContrller = nil;
}
/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: Sources/SSAlertSwift/SSAlertView.swift
//
//  SSAlertView.swift
//  SSAlert
//
//  Created by yangsq on 2021/8/13.
//

import UIKit

extension SSAlertView {
    public enum BackgroundMaskType {
        /// 默认没有遮罩
        case none
        
        /// 透明遮罩
        case clear
        
        /// 黑色遮罩
        case black
    }
}

fileprivate extension SSAlertView {
    typealias  HideCompletion = () -> Void
    class HideCompletionData: NSObject {
        var completion: HideCompletion?
    }
}

open class SSAlertView: UIView {
    /// customView 的内间距
    public var customEdgeInsets: UIEdgeInsets = .zero {
        didSet {
            setNeedsDisplay()
            layoutIfNeeded()
        }
    }
    
    /// 展示和隐藏动画设置
    public var animation: SSAlertAnimation
    
    ///遮罩视图
    public private(set) var backgroundMask: UIView?
    
    /// 是否点击遮罩隐藏，默认点击会隐藏
    public var canTouchMaskHide: Bool = true
    /// 点击遮罩隐藏是否开启动画，默认nil,跟随展示的时候设置是否开启动画
    public var isTouchMaskHideAnimated: Bool? = nil
    
    /// 模态视图弹窗才有
    public weak private(set) var navigationController: UINavigationController?
    
    public var isHideStatusBar = false {
        didSet {
            self.toViewContrller?.isHideStatusBar = isHideStatusBar
        }
    }
    public var isShowNavWhenViewWillDisappear = true {
        didSet {
            self.toViewContrller?.isShowNavWhenViewWillDisappear = isShowNavWhenViewWillDisappear
        }
    }
    private var onView: UIView
    public private(set)  var maskType: BackgroundMaskType
    public private(set) var customView: UIView
    private weak var fromViewController: UIViewController?
    private weak var toViewContrller: SSAlertAnimationController?
    private var presentAnimation: SSAlertPresentAnimation?
    private var canPanDimiss: Bool = false
    private var hideCompletions = [HideCompletionData]()
    /// 初始化（普通视图弹窗）
    public init(customView: UIView,
                onView: UIView,
                animation: SSAlertAnimation = SSAlertDefaultAnmation(state: .fromCenter),
                maskType: BackgroundMaskType = .black) {
        self.animation = animation
        self.onView = onView
        self.maskType = maskType
        self.customView = customView
        super.init(frame: .zero)
        makeUI()
    }
    
    /// 初始化（模态视图弹窗),canPanDimiss: 是否支持拖拽消息
    public convenience init(customView: UIView,
                            fromViewController: UIViewController,
                            animation: SSAlertAnimation = SSAlertDefaultAnmation(state: .fromCenter),
                            navigationControllerClass: UINavigationController.Type = UINavigationController.self,
                            maskType: BackgroundMaskType = .black,
                            canPanDimiss: Bool = true,
                            canTouchMaskHide: Bool = true) {
        let animaionVC = SSAlertAnimationController()
        self.init(customView: customView, onView: animaionVC.view, animation: animation, maskType: maskType)
        animaionVC.isHideStatusBar = self.isHideStatusBar
        if let superViewController = superViewController(view: customView), superViewController != animaionVC {
            animaionVC.addChild(superViewController)
            superViewController.didMove(toParent: animaionVC)
        }
        self.canPanDimiss = canPanDimiss
        self.canTouchMaskHide = canTouchMaskHide
        self.fromViewController = fromViewController
        let nav = navigationControllerClass.init(rootViewController: animaionVC)
        nav.modalPresentationStyle = .custom
        self.navigationController = nav
        self.canPanDimiss = canPanDimiss
        let animation = SSAlertPresentAnimation(navigationViewController: nav, animationView: customView, canPanDimiss: canPanDimiss)
        self.navigationController?.transitioningDelegate = animation
        self.navigationController?.modalPresentationCapturesStatusBarAppearance = true
        self.presentAnimation = animation
        self.toViewContrller = animaionVC
        
    }
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        backgroundMask?.frame = CGRect(x: 0, y: 0, width: onView.ss_w, height: onView.ss_h)
        customView.ss_x = customEdgeInsets.left
        customView.ss_y = customEdgeInsets.top
        customView.ss_w = frame.width - customEdgeInsets.left - customEdgeInsets.right
        customView.ss_h = frame.height - customEdgeInsets.top - customEdgeInsets.bottom
    }
    
    
    private func makeUI() {
        if maskType != .none {
            let backgroundMask = UIButton()
            backgroundMask.alpha = 0
            onView.addSubview(backgroundMask)
            backgroundMask.addTarget(self, action: #selector(backgroundMaskAction), for: .touchUpInside)
            self.backgroundMask = backgroundMask
            if maskType == .black {
                backgroundMask.backgroundColor = UIColor.black.withAlphaComponent(0.3)
            }
        }
        addSubview(customView)
        onView.addSubview(self)
    }
    /// 展示
    public func show(animated: Bool = true, completion: (() -> Void)? = nil) {
        if navigationController != nil {
            presentView(animated: animated, completion: completion)
        } else {
            showView(animated: animated, completion: completion)
        }
    }
    /// 隐藏
    public func hide(animated: Bool = true, completion: (() -> Void)? = nil) {
        if navigationController != nil {
            dimissView(animated: animated, completion: completion)
        } else {
            hideView(animated: animated, completion: completion)
        }
    }
    
    /// 展示（普通视图弹窗）
    /// - Parameter animated: 是否动画
    private func showView(animated: Bool = true, completion: (() -> Void)? = nil) {
        isTouchMaskHideAnimated = animated
        var customSize = customView.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize)
        if customView.ss_size != .zero {
            customSize = customView.ss_size
        }
        let size = CGSize(width: customSize.width + customEdgeInsets.left + customEdgeInsets.right, height: customSize.height + customEdgeInsets.top + customEdgeInsets.bottom)
        animation.showAnimationOfAnimationView(animationView: self, viewSize: size, animated: animated, completion: { _ in completion?() })
    }
    /// 隐藏（普通视图弹窗）
    /// - Parameter animated: 是否动画
    private func hideView(animated: Bool = true, completion: (() -> Void)? = nil) {
        var customSize = customView.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize)
        if customView.ss_size != .zero {
            customSize = customView.ss_size
        }
        let size = CGSize(width: customSize.width + customEdgeInsets.left + customEdgeInsets.right, height: customSize.height + customEdgeInsets.top + customEdgeInsets.bottom)
        animation.hideAnimationOfAnimationView(animationView: self, viewSize: size, animated: animated) { [weak self] _ in guard let self = self else {  return false }
            completion?()
            
            self.hideCompletions.forEach { hideC in
                hideC.completion?()
                hideC.completion = nil
                self.hideCompletions.removeAll(where: {$0 == hideC})
            }
            
            return false
        }
        
    }
    /// 展示（模态视图弹窗）
    /// - Parameter animated: 是否动画
    private func presentView(animated: Bool = true, completion: (() -> Void)? = nil) {
        isTouchMaskHideAnimated = animated
        var customSize = customView.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize)
        if customView.ss_size != .zero {
            customSize = customView.ss_size
        }
        let size = CGSize(width: customSize.width + customEdgeInsets.left + customEdgeInsets.right, height: customSize.height + customEdgeInsets.top + customEdgeInsets.bottom)
        presentAnimation?.animationCompletion(showAnimation: {
            [weak self] in guard let self = self else { return }
            self.animation.showAnimationOfAnimationView(animationView: self, viewSize: size, animated: animated) { _ in
                self.presentAnimation?.setCompleteTransitionIsHide(isHide: false)
                completion?()
            }
        })
        self.presentAnimation?.duration = animation.animationDuration()
        ///存在的问题： app当前显示的是弹窗界面，默认presnent出来的视图且modalPresentationStyle不是CurrentContext这这种，这个视图的presentingVC默认根控制器，例如存在tabbarVC的话就是tabbarVC，这个会造成一个问题：当这个视图dimiss的时候，会把所有的presentedVC都dimiss（其实调用dimiss的时候是调用当前视图的presentigngVC去dimiss,遵循随创建随释放）
        /// 这个通过设置 definesPresentationContext = true,modalPresentationStyle = UIModalPresentationOverCurrentContext，这样当前视图的presentigngVC为 self.navigationController,
        /// UIModalPresentationOverCurrentContext和 UIModalPresentationCurrentContext的区别是弹出的界面，前者不会把底部的视图stact移除，后者则会。
        self.navigationController?.definesPresentationContext = true
        self.navigationController?.modalPresentationStyle = .overCurrentContext
        self.fromViewController?.present(self.navigationController!, animated: true, completion: nil)
        if canPanDimiss {
            dimissViewDragToHideAnimation(animated: animated)
        }
    }
    /// 隐藏（模态视图弹窗）
    /// - Parameter animated: 是否动画
    private func dimissView(animated: Bool = true, completion: (() -> Void)? = nil) {
        var customSize = customView.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize)
        if customView.ss_size != .zero {
            customSize = customView.ss_size
        }
        let size = CGSize(width: customSize.width + customEdgeInsets.left + customEdgeInsets.right, height: customSize.height + customEdgeInsets.top + customEdgeInsets.bottom)
        self.presentAnimation?.duration = animation.animationDuration()
        /// 没动画时，不会走 presentAnimation里面的动画
        if !animated {
            self.toViewContrller?.dismiss(animated: animated, completion: nil)
            self.animation.hideAnimationOfAnimationView(animationView: self, viewSize: size, animated: animated) { [weak self] _ in guard let self = self else { return true }
                self.handleHideAction(completion: completion)
                return false
            }
        } else {
            presentAnimation?.animationCompletion(hideAnimation: {
                [weak self] in guard let self = self else { return }
                self.animation.hideAnimationOfAnimationView(animationView: self, viewSize: size, animated: animated) { [weak self] _ in guard let self = self else { return true }
                    return self.presentAnimation!.setCompleteTransitionIsHide(isHide: true)
                }
            })
            self.toViewContrller?.dismiss(animated: animated, completion: {
                [weak self] in guard let self = self else { return }
                self.handleHideAction(completion: completion)
            })
        }
    }
    
    private func dimissViewDragToHideAnimation(animated: Bool) {
        var customSize = customView.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize)
        if customView.ss_size != .zero {
            customSize = customView.ss_size
        }
        let size = CGSize(width: customSize.width + customEdgeInsets.left + customEdgeInsets.right, height: customSize.height + customEdgeInsets.top + customEdgeInsets.bottom)
        presentAnimation?.animationCompletion(hideAnimation: {
            [weak self] in guard let self = self else { return }
            self.animation.hideAnimationOfAnimationView(animationView: self, viewSize: size, animated: animated) { _ in
                return self.presentAnimation!.setCompleteTransitionIsHide(isHide: true)
            }
        },endCompletion: {
            [weak self] in guard let self = self else { return }
            self.handleHideAction()
        }, panDimissAnimation: { [weak self] point, frame in
            guard let self = self else { return (0, 0.4) }
            return self.animation.panToDimissTransilatePoint(point: point, panViewFrame: frame)
        }, dimissAnimation: { [weak self] isCancel in guard let self = self else { return }
            self.toViewContrller?.isDimiss = !isCancel
        })
        self.presentAnimation?.duration = animation.animationDuration()
    }
    
    public func refreshFrame(animated: Bool = true) {
        var customSize = customView.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize)
        if customView.ss_size != .zero {
            customSize = customView.ss_size
        }
        let size = CGSize(width: customSize.width + customEdgeInsets.left + customEdgeInsets.right, height: customSize.height + customEdgeInsets.top + customEdgeInsets.bottom)
        animation.refreshAnimationOfAnimationView(animationView: self, viewSize: size, animated: animated, completion: nil)
    }
    private func handleHideAction(completion: (() -> Void)? = nil) {
        completion?()
        self.hideCompletions.forEach { hideC in
            hideC.completion?()
            hideC.completion = nil
            self.hideCompletions.removeAll(where: {$0 == hideC})
        }
    }
    
    @objc private func backgroundMaskAction() {
        if canTouchMaskHide {
            if fromViewController != nil {
                dimissView(animated: isTouchMaskHideAnimated ?? false)
            } else {
                hideView(animated: isTouchMaskHideAnimated ?? false)
            }
        }
    }
    
    private func superViewController(view: UIView) -> UIViewController?  {
        var next = view.next
        while next != nil {
            if next is UIViewController {
                return next as? UIViewController
            } else {
                next = next!.next
            }
        }
        return nil
    }
    
    public func observeHideCompletion(completion: (() -> Void)?) {
        let completionData = HideCompletionData()
        completionData.completion = completion
        self.hideCompletions.append(completionData)
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}






